docs/enhanced-provider-k8s.md

```diff
@@ -0,0 +1,303 @@
+
+### About The Demo Project
+
+The project `capabilities-shim` is aimed to demonstrate the composing of software capabilities from different applications to build user-oriented solution. It leverages the powerful Crossplane composition engine, the Provider Kubernetes, along with other cloud native install methods run at back end such as Operator Lifecycle Manager (OLM) to provision the actual workloads in a declarative way.
+
+The project defined a couple of sample capabilities that can be installed:
+
+* `Networking` capability to privion a Kong proxy as an API Gateway.
+* `Logging` capability to provision an Elasticsearch and Kibana instance for logging storage and visualization.
+* `All` capability to compose all the above capabilities as an all-in-one delivery.
+
+![](images/architecture.png)
+
+In this document, I will use `capabilities-shim` with the help of Provider Kubernetes and OLM to provision `Networking` capability.
+
+Firstly, let's run below command to launch the demo environment:
+
+```console
+demo/install.sh
+```
+
+Then, run below command to install `Networking` capability:
+
+```console
+kubectl apply -f demo/examples/networking/olm.yaml
+```
+
+Before we use the enhanced provider, this will launch both the Kong operator pod and the Kong pod, also all managed resources handled by Provider Kubernetes and external resources handled by OLM:
+
+* For managed resources, there are two `Object` resources where one represents the `Subscription` resource and the other one represents the `Kong` resource. The creation of managed resources will trigger the creation of external resources which is handled by Provider Kubernetes.
+* For external resources, as the `Subscription` resource is created, it triggers the creation of the `ClusterServiceVersion` resource, then the `InstallPlan` resource, which triggers the launch of Kong operator pod. Then, the creation of `Kong` resource will be detected by Kong operator pod and ultimately triggers the launch of Kong pod.
+
+![](images/default-launch.png)
+
+Everything works perfect, isn't? Now let's go through the uninstall process and see where the problems are.
+
+### Unmanaged External Ressource
+
+To uninstall the capability can be done by simply deleting the corresponding Composite Resource Claim(XRC) for the capability:
+
+```console
+kubectl delete networkingclaim my-networking-stack
+```
+
+This triggers the deletion of managed resources, i.e. the `Object` resources, then the deletion of external resources, in this case, the `Subscription` and `Kong` resource. This triggers the deletion of the Kong pod since `Kong` resource is deleted.
+
+However, as the `ClusterServiceVersion` resource still exists, so does the `InstallPlan` resource, the Kong operator pod still remains. This is considered as not a clean uninstall since some garbage resources remain in the cluster.
```

Comment:
Agreed. This might not be a normal case. I have the same feeling when I see this issue for the first time. But this appears to be the expected behavior per https://github.com/operator-framework/operator-lifecycle-manager/issues/1168. They want to have an option that the operator can still run even after `Subscription` is deleted, and to delete the operator, user has to delete `ClusterServiceVersion` explicitly.

For those enumerable values of resource management policy, it is just the results after I explore all combinations for provider operations, i.e. CRUD, even that I have combined C + U as a single operation. But for sure `Undeletable` and `Observable` are more often used, while `ObservableAndDeletable` are less often used.

Reply:
I see, yeah, having a less used option shouldn't hurt. 

However, I am wondering if we could find better names for the policies. Honestly, I am struggling to understand what a `ObservableAndDeletable` means without checking the table. Given this would be our API, it worth trying our best to make it as explicit as possible :) 

I think the biggest flaw with current naming is some implies what policy `could do` (Observable, ObservableAndDeletable) and some `could not do` (Undeletable). 

After quick thinking, I believe we can just use whatever policy is allowed to do:

ObservableAndDeletable => ObserveAndDelete
Undeletable => ObserveAndUpdate (Create would be implicit here)
Observable => Observe

or sticking with CRUD:

ObservableAndDeletable => ReadDelete
Undeletable => ReadCreateUpdate
Observable => Read

What do you think?

Reply:
Updated the doc according to the proposed policy names. I like the first option, as it exactly maps to the Observe, Create, Update, Delete operations defined in Crossplane Provider. Just made a very little change as below since it makes the mapping more obviously:

ObservableAndDeletable => ObserveDelete
Undeletable => ObserveCreateUpdate
Observable => Observe